Add aggregation functionality for #GROUP_BY function

lbschanno opened the pull request:
Provide the ability to aggregate fields when grouping. Specifically, add the new functions #SUM, #MIN, #MAX, #COUNT, #AVERAGE that will determine the aggregate value for any specified fields against the entries that match any particular grouping.

Resolves #1687

lbschanno committed:
Add aggregation functionality for #GROUP_BY function

Provide the ability to aggregate fields when grouping. Specifically, add
the new functions #SUM, #MIN, #MAX, #COUNT, #AVERAGE that will determine
the aggregate value for any specified fields against the entries that
match any particular grouping.

Resolves #1687

lbschanno committed:
Fix failing unit tests

avgAGB commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/GroupingAttribute.java:
```diff
@@ -0,0 +1,58 @@
+    }
+    
+    /**
+     * Returns a hashcode of the attribute's value.
```
Really small thing, could you add a @return notation for the hashCode javadoc?

lbschanno replied:
Not a problem, I just added a bunch of documentation that included this change.

lbschanno committed:
Add documentation

lbschanno committed:
Rename methods for clarity

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/AggregatedFields.java:
```diff
@@ -0,0 +1,312 @@
+ */
+public class AggregatedFields {
+    
+    private final Multimap<String,Aggregator<?>> aggregatorMap;
```
Thoughts on updating this to be Map<String, Map<String, Aggregator>> keyed by both the field, and the aggregator enum name?  Might make the merge operation faster.

lbschanno replied:
Agreed, I'll see what I can do here to improve it.

lbschanno replied:
Done.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/AggregatedFields.java:
```diff
@@ -0,0 +1,312 @@
+         * @return a new {@link Factory} instance with deconstructed fields
+         */
+        public Factory deconstructIdentifiers() {
+            Set<String> sumFields = this.sumFields.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+            Set<String> maxFields = this.maxFields.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+            Set<String> minFields = this.minFields.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+            Set<String> countFields = this.countFields.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+            Set<String> averageFields = this.averageFields.stream().map(JexlASTHelper::deconstructIdentifier).collect(Collectors.toSet());
+            return new Factory().withSumFields(sumFields).withMaxFields(maxFields).withMinFields(minFields).withCountFields(countFields)
+                            .withAverageFields(averageFields);
```
Do we not want to deconstruct the identifiers when fields are added in the first place?

lbschanno replied:
We should, I'll make the change.

lbschanno replied:
Done.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/AverageAggregator.java:
```diff
@@ -0,0 +1,151 @@
+    }
+    
+    @Override
+    public void merge(Aggregator<?> other) {
```
While reading this code, I had a brief concern about the potential for concurrent calls to merge in the pipeline iterator.  I know that we are currently not running with more than one evaluation pipeline, but if we did, would this code still work?  Actually, would grouping work at all?  Are we testing this somewhere already?

lbschanno replied:
As far as I know, we're not testing grouping with concurrent calls.

ivakegg replied:
The grouping is being done after the pipeline iterator so no need for thread safety here.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/CountAggregator.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * The total number of times the field was seen.
+     */
+    private int count;
```
int, really?

lbschanno replied:
Can you please elaborate on your specific concerns and what you would like to see here instead?

jwomeara replied:
just wondering if we ever expect to overrun an int while counting.  in that case, a long might be a better choice.  i'll leave that decision to you and Ivan.

lbschanno replied:
It's not an unreasonable concern to have, and I don't mind making it a long to help safeguard against overruns.

lbschanno replied:
Done.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/common/grouping/Groups.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * A map of distinct grouping values to their groups.
+     */
+    private final Map<Set<GroupingAttribute<?>>,Group> groups = new HashMap<>();
```
A Set seems like an expensive object to use as a key in a hashmap.  Each time it calls hashCode on the set (at a minimum every time we call `get` on the hashmap), it iterates through every key value pair in the set and computes the hashcode.

lbschanno replied:
This is something I've also wanted to improve upon, it's how the original code handled finding groupings. I'll work on coming up with a more efficient key object to use instead of a set.

jwomeara replied:
I approved, but I just want to double check that we are going to proceed with a set as the key for this map.  Just wondering where your thought process lead you with this one.

lbschanno replied:
So I found myself going back and forth on this, trying to figure out a workable solution that wasn't effectively just a wrapper around a Set. One approach we could try is an immutable HashSet implementation that computes the hashcode once at instantiation, retaining that as an attribute to be returned from hashCode(), and using that as the map key. Thoughts?

jwomeara replied:
I am ok with whatever you decide - i just wanted to hear your thoughts.  If you're ok with it as-is, then we can move forward with that.

lbschanno replied:
I think I'm okay with it as is, but I also don't really have any data on how large these grouping sets can get, nor how large the map keyset can get either, so I can't really say if this optimization is necessary or not. If we want to take a safer approach, we should try to optimize it. If we want to wait and see, it can stay as is.

lbschanno replied:
@jwomeara I pushed a commit that optimizes the hashcode() usage, please take a look if you get a chance and let me know what you think.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/iterator/GroupingIterator.java:
```diff
@@ -71,27 +75,17 @@ public class GroupingIterator implements Iterator<Map.Entry<Key,Document>> {
-    public GroupingIterator(Iterator<Map.Entry<Key,Document>> previousIterators, MarkingFunctions markingFunctions, Collection<String> groupFieldsSet,
-                    int groupFieldsBatchSize, YieldCallback<Key> yieldCallback, long resultTimeout) {
+    public GroupingIterator(Iterator<Map.Entry<Key,Document>> previousIterators, MarkingFunctions markingFunctions, Collection<String> groupFields,
+                    int groupFieldsBatchSize, YieldCallback<Key> yieldCallback, AggregatedFields.Factory aggregateFieldsFactory) {
```
Did we lose the functionality that utilizes the resultTimeout?

lbschanno replied:
It looks like resultTimeout was added in commit 3b7a68004289971232b7833206156db5b94fa2e4 by Ivan, but no functionality was added that actually used it. 

@ivakegg, was there supposed to be functionality added to GroupingIterator for resultTimeout?

ivakegg replied:
There is actually a story behind that one.  The grouping mechanism essentially cannot return results to the user until all results on the back end have been gathered.  We were working the ability to allow the iterator stack to return all the way back to the RunningQuery which could then return an empty page to the user and subsequently continue the query with a new next call.  This would prevent the query from timing out.  So to that end this timeout was supposed to work in concert with the timeout in the grouping transform which is used webserver side.
  You will see that the timeout mechanism was actually implemented in the GroupingTransform which is used webserver side.  Then we realized that the RunningQuery could essentially just run the back end "next" call in a separate thread and manage the returning of an empty results page back to the user without having to modify the grouping logic.
  The timeout being passed to the GroupingIterator is a leftover of that.  I was going to go back and actually use that to allow the iterator stack to return and be torn down and rebuilt to avoid scan thread starvation.  However Bill is working that problem in a different way so that parameter can be removed.

lbschanno replied:
Thanks for the background info, I'll leave the parameter out.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/language/builder/jexl/JexlTreeBuilder.java:
```diff
@@ -63,7 +68,7 @@ public class JexlTreeBuilder extends QueryTreeBuilder {
             new MatchesAtLeastCountOf(), new Jexl(), new TimeFunction(), new AtomValuesMatchFunction(), new Options(), new Unique(), new GroupBy(),
             new EvaluationOnly(), new UniqueByDay(), new UniqueByHour(), new UniqueByTenthOfHour(), new UniqueByMonth(), new UniqueByYear(),
-            new UniqueByMinute(), new NoExpansion(), new Compare(), new ExcerptFields()};
+            new UniqueByMinute(), new NoExpansion(), new Compare(), new ExcerptFields(), new Sum(), new Min(), new Max(), new Average(), new Count()};
```
These should be added to allowedQueryFunctions in QueryLogicFactory.xml as well.

lbschanno replied:
Will do.

lbschanno replied:
After double-checking, these were already added to allowedQueryFunctions in QueryLogicFactory.xml.

jwomeara replied:
I noticed that as well and forgot to delete this comment.  My apologies.

lbschanno replied:
No worries.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:
```diff
@@ -2049,7 +2143,11 @@ public Set<String> getOptionalQueryParameters() {
         optionalParams.add(datawave.webservice.query.QueryParameters.QUERY_PAGESIZE);
         optionalParams.add(datawave.webservice.query.QueryParameters.QUERY_PAGETIMEOUT);
         optionalParams.add(datawave.webservice.query.QueryParameters.QUERY_EXPIRATION);
-        optionalParams.add(datawave.webservice.query.QueryParameters.QUERY_MAX_RESULTS_OVERRIDE);
```
Was this removal intentional?

lbschanno replied:
It was not, I think this got missed when resolving some conflicts with updates from integration. I'll put it back in.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/transformer/GroupingTransform.java:
This class seems to share some code with GroupingIterator.  Any chance we can identify the common code and pull it out to a common location?

lbschanno replied:
No problem, I'll do so.

jwomeara commented on warehouse/query-core/src/test/java/datawave/query/transformer/GroupingTest.java:
Lots of changes in here - I admit that I only skimmed.  Did we remove the teardown aspects of these tests though?

lbschanno replied:
No, the teardowns and interrupts are still used. The function `collectQueryResults()` will run the query under test against each permutation of the teardown/interrupts and collect the results for validations. I simplified the creation of the set of teardowns and interrupts via `EnumSet.allOf()`.

jwomeara reviewed (changes requested):
Overall, looks really good.  Just a few comments throughout.  

There were a few places where it seemed like the logic Ceria added to return a blank page after a timeout might have been lost, but maybe it was just moved and I had trouble tracking.

One overall design thought - do we want to be able to impose limits on which fields can be aggregated?  A configurable allow list perhaps?

Because of how extensive these changes are, we should really put this through it's paces before merging IMO.

lbschanno committed:
Add back accidentally-removed query parameter

lbschanno commented:
@jwomeara 
> One overall design thought - do we want to be able to impose limits on which fields can be aggregated? A configurable allow list perhaps?

I think @ivakegg will need to be the one to decide if that's a requirement for this work. If so, it should be relatively simple to implement.

ivakegg commented:
> 

If there are any restriction on fields to be grouped on or fields to be summed/aggregated, that would be done by the webserver.  Currently there are no restrictions as such and no requirements for any.

lbschanno committed:
Major edits to grouping and aggregation logic

Make major changes to how grouping and aggregation is handled based on
whether or not the field of an event has a grouping context and
instance.

Additionally, add GroupAggregateFields to act as a central place to
collect the configured group and aggregate fields supplied by the user.

lbschanno committed:
Delete unused imports, classes, and methods

lbschanno committed:
Merge branch 'integration' into feature/aggregateGroupedFields

lbschanno committed:
Add back accidentally deleted tests

lbschanno committed:
Improve documentation

lbschanno committed:
Minor edits

lbschanno committed:
Fix failing tests

lbschanno committed:
Merge branch 'integration' into feature/aggregateGroupedFields

lbschanno committed:
Make GroupFields serializable

lbschanno committed:
Merge branch 'integration' into feature/aggregateGroupedFields

lbschanno committed:
Code formatting from build

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into feature/aggregateGroupedFields

lbschanno committed:
Merge branch 'integration' into feature/aggregateGroupedFields

ivakegg committed:
Merge branch 'integration' into feature/aggregateGroupedFields

jwomeara reviewed (dismissed):
One minor comment left unresolved, but otherwise looks great!

lbschanno committed:
Optimize hashcode() usage for grouping keys

lbschanno committed:
Improve documentation

lbschanno committed:
Fix code formatting

ivakegg committed:
Merge branch 'integration' into feature/aggregateGroupedFields

ivakegg committed:
Merge branch 'integration' into feature/aggregateGroupedFields

ivakegg committed:
Updated to specify a reverse model mapping separate from the
inverseModelMap to ensure unique reverse mappings

ivakegg committed:
Updated to handle the reverse model mapping changes

lbschanno committed:
Pare reverseModel down to relevant mappings only

lbschanno committed:
Redefine what counts as a valid grouping

ivakegg committed:
Reduce the fields to be the mapped set of fields and avoid remapping the
fields everywhere that we do not need to.

ivakegg committed:
Merge branch 'integration' into feature/aggregateGroupedFields

ivakegg committed:
Merge branch 'integration' into feature/aggregateGroupedFields
